corehq/apps/app_manager/models.py

```diff
@@ -5577,8 +5579,9 @@ class LinkedApplication(Application):
     """
     An app that can pull changes from an app in a different domain.
     """
-    # This is the id of the master application
-    master = StringProperty()
+    master = StringProperty()   # Legacy, should be removed once all linked apps support multiple masters
+    upstream_app_id = StringProperty()      # ID of the app that was most recently pulled
```

Comment:
What's the difference between master and upstream? When should we say master vs upstream or will they be used interchangeably?

Reply:
Jenny and I have been discussing this.  Here's my take on the terminology:

In the current world each linked app has a single "master" app.  In the future world, each linked app will be able to pull from multiple "upstream" apps, all of which descend from a single "progenitor" app.

What do you think?

Reply:
Interesting. Are there going to be any constraints on enforcing the same "progenitor" app? Does that mean that master as a linked app term will never be used after this PR migrates current linked apps?

I'm mostly asking for my own understanding as we've overloaded "master" when talking about this in the past. As long as we're clear on the terminology being used, I'll probably be happy with the outcome

Reply:
> Does that mean that master as a linked app term will never be used after this PR migrates current linked apps?

Yeah, that's the idea.  The "master" terminology is for sure ambiguous (`master_id` is an unrelated concept on the same model), so hopefully this'll clear it up.

Reply:
I'd like to wholly get rid of "master" as it relates to linked apps, but that's a rather large change and would also affect linked domains. My major complaint is the one Ethan mentions: in current production, `VersionedDoc` has a `master_id` property, `LinkedApplication` has a `master` property, and while both of these are strings that contain an app id, they have nothing to do with each other. Come to think of it, it might be easier to rename `master_id` (not sure to what...`canonical_id`?) and let linked apps/projects keep "master," rather than introducing "upstream" - @esoergel what do you think?